Semantic world now handles recognized obects and pushes them into planning scene

sachinchitta opened the pull request:
(1) added more options for pick and place, (2) added place exploration to the gui

sachinchitta committed:
reading back state from planning interface to identify tables at startup, adding more options for retreat motions for pick and approach motions for place

sachinchitta committed:
removing commented code

Commit:
adding more options for retreat for pick and approach for place

Commit:
Merge branch 'sachin-hydro' of github.com:ros-planning/moveit_ros into sachin-hydro

sachinchitta committed:
Merge branch 'sachin-hydro' of github.com:ros-planning/moveit_ros into sachin-hydro

sachinchitta committed:
letting semantic world handle recognized objects as well

sachinchitta committed:
resetting index when recognized objects are cleared, new default vertical distance

sachinchitta committed:
adding place exploration to gui

sachinchitta committed:
Merge branch 'hydro-devel' into sachin-hydro

isucan commented on manipulation/move_group_pick_place_capability/src/pick_place_action_capability.cpp:
```diff
@@ -457,7 +458,31 @@ void move_group::MoveGroupPickPlaceAction::fillGrasps(moveit_msgs::PickupGoal& g
-      goal.possible_grasps = response.grasps;
+    {
+      random_numbers::RandomNumberGenerator rng;
+      unsigned int num_retreat_directions = 10;
```
This needs to be a ros param. What does it do, actually?

isucan commented on manipulation/move_group_pick_place_capability/src/pick_place_action_capability.cpp:
```diff
@@ -457,7 +458,31 @@ void move_group::MoveGroupPickPlaceAction::fillGrasps(moveit_msgs::PickupGoal& g
     if (grasp_planning_service_.call(request, response))
-      goal.possible_grasps = response.grasps;
+    {
+      random_numbers::RandomNumberGenerator rng;
```
This is expensive to construct. Please make it a member variable.

isucan commented on manipulation/move_group_pick_place_capability/src/pick_place_action_capability.cpp:
```diff
@@ -457,7 +458,31 @@ void move_group::MoveGroupPickPlaceAction::fillGrasps(moveit_msgs::PickupGoal& g
+    {
+      random_numbers::RandomNumberGenerator rng;
+      unsigned int num_retreat_directions = 10;
+      double max_pitch_angle = M_PI/10;
```
please use boost::math::constants::pi<double>() instead of M_PI

isucan commented on manipulation/move_group_pick_place_capability/src/pick_place_action_capability.cpp:
```diff
@@ -457,7 +458,31 @@ void move_group::MoveGroupPickPlaceAction::fillGrasps(moveit_msgs::PickupGoal& g
+      for(std::size_t i = 0; i < response.grasps.size(); ++i)
+      {
+	manipulation_msgs::Grasp grasp = response.grasps[i];
+	for(std::size_t j = 0; j < num_retreat_directions; ++j)
```
This is very magic behaviour that the user is not asking for. I am reluctant to merge this.

sachinchitta replied:
You are right - this is probably the wrong place for this and the wrong way to implement this. The issue is the following: a Grasp contains an approach and retreat direction. However, there may be other retreat directions (for the pickup) and other approach directions (for the place) that could be useful in pick and place, e.g. straight up or down works pretty well a lot of times. It would be useful to specify these additional directions as well in the messages. I propose adding a vector of retreat directions into the PickAction and a vector of approach directions into the PlaceAction. These would be added to the current default directions provided by the grasping node.

isucan replied:
This is a good idea. I think this should go in the message. The constants that are used in the code could roughly become members of the message. Before we do this change though, I'd like to include the change in the representation of gripper commands, using the newly merged effort[] member in common msgs. Would this be ok?

sachinchitta replied:
Yup, that sounds fine.

mikeferguson replied:
Should approach/retreat both be vectors in both the Pick and Place? Simply for a consistent looking interface, rather than having one with a vector of approach and single retreat, and the other with a vector of retreat and a single approach?

isucan replied:
@mikeferguson yes, that is the plan (although my message above does not seem to say that :) ). We will have vectors of directions instead of just direction. I am still debating whether we should leave the simple direction in the message in for compatibility, but I think I'd rather break things a bit.

mikeferguson replied:
@isucan, given that moveit is still in development, I'd prefer to avoid bogging it down with compatibility already. As usual, I think the most important thing is to announce the change to the mailing list. I presume  in his comment below, @davetcoleman is mainly concerned about another "change in how place locations are interpreted", which really wasn't announced to the mailing list (or anywhere much for that matter), which I know caused quite a bit of frustration for both of us when demos suddenly started failing in non-obvious ways.

isucan commented on perception/semantic_world/include/moveit/semantic_world/semantic_world.h:
```diff
@@ -59,7 +59,7 @@ class SemanticWorld
 
   /** @brief The signature for a callback on receiving table messages*/
-  typedef boost::function<void()> TableCallbackFn;
+  typedef boost::function<void()> SemanticWorldCallbackFn;
```
is this still just for table messages?

sachinchitta replied:
No - also used for recognized objects now

isucan commented on planning_interface/CMakeLists.txt:
```diff
@@ -50,6 +51,7 @@ catkin_package(
     moveit_ros_warehouse 
     moveit_ros_manipulation
     moveit_ros_move_group
+    random_numbers
```
This is not exposed in a header file. We do not need it here.

isucan commented on planning_interface/move_group_interface/src/move_group.cpp:
```diff
@@ -371,22 +372,44 @@ class MoveGroup::MoveGroupImpl
+      //Add a few random directions in addition to the straight down direction
+      Eigen::Vector3d approach_direction, down(-Eigen::Vector3d::UnitZ());
+      unsigned int num_approach_directions = 10;
+      double max_pitch_angle = M_PI/4.0;
```
I do not understand this. This looks like a hack....

davetcoleman commented:
I'm actively using the pick and place pipeline so please announce any major changes to this that might affect my code, thanks!

sachinchitta committed:
moved sampling options for approach and retreat directions out of pick and place so that they are user driven

sachinchitta committed:
changed API to make orientations sampling for place a parameter as well, most parameters now specified by user through GUI

Commit:
changed INFO's to DEBUG

Commit:
Merge branch 'sachin-hydro' of github.com:ros-planning/moveit_ros into sachin-hydro

Commit:
Merge branch 'hydro-devel' into sachin-hydro

sachinchitta commented:
I updated the pull request and moved hard-coded stuff out of the planning interface and into the GUI. I also added a pickup directions to the PickupAction. The PlaceLocations msg already allows you to specify place locations and directions independent of the grasp action.

isucan commented:
This can't be merged now and messages have changed a bit since this pull request was done. We can reopen later on.
